### ISSUE TITLE
`impl AsyncBufRead for Decoder`

### ISSUE THREAD

**Issue by DrewMcArthur** [Nullus157/async-compression#289]
Hi! So I've got a `stream: impl AsyncBufRead`, which may or may not be gzipped.  I'm trying to return that stream in a function whose return type is `Result<impl AsyncBufRead, Error>`, but running into the issue where the `GzipDecoder` is `AsyncRead`, but not `AsyncBufRead`.  Since the `inner` reader it pulls from is `AsyncBufRead`, is there any reason why the decoder itself can't be `AsyncBufRead`?

Or, am I misreading the docs, and I should just be calling `into_inner()`? the way I read it currently, this returns the original stream though, not the decoded version.

If it's easier, I just need this for the `GzipDecoder`, but I assume this level would be generic across all the `Decoder`s.

**Comment by DrewMcArthur** [comment]
update: I was able to update my function to return a `Box<dyn AsyncRead>`, which covers both the GZipped case (returning a `GzipDecoder: AsyncRead`) and the plain filestream from S3 (returning an `impl AsyncBufRead`, which also is `AsyncRead`).  Since my parsing function downstream just needs a `Stream<Item=Bytes>`, that's all I needed.

I'm still very curious what it would take to implement `AsyncBufRead` for the `Decoder`.  I tried messing around with it a bit last night, but I'm too unfamiliar with `Pin`s and all of that.
